File: Data Structures & Algorithms/course-schedule/submission-1.java
class Solution {
    HashMap<Integer, List<Integer>> adj = new HashMap<>();
    Set<Integer> visiting = new HashSet<>();
    public boolean canFinish(int numCourses, int[][] prerequisites) {

        for (int i = 0; i < prerequisites.length; i++) {
            int c = prerequisites[i][0];
            int p = prerequisites[i][1];
            List<Integer> n = adj.get(c);
            if (n == null) {
                List<Integer> l = new ArrayList<>();
                l.add(p);
                adj.put(c, l);
            } else {
                n.add(p);
            }
        }

        for (int i : adj.keySet()) {
            if (!dfs(i)) {
                return false;
            };
        }
        return true;
    }

    boolean dfs(int course) {
        if (adj.get(course) == null || adj.get(course).size() == 0) {
            return true;
        }
        if (visiting.contains(course)) {
            return false;
        }
        visiting.add(course);
        for (int prereq : adj.get(course)) {
            if (!dfs(prereq)) {
                return false;
            }
        }
        visiting.remove(course);
        adj.put(course, null);
        return true;
    }
}

File: Data Structures & Algorithms/course-schedule-ii/submission-1.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        self.adj = {i: [] for i in range(numCourses)}
        for p in prerequisites:
            self.adj[p[0]].append(p[1])
        self.order = []
        for c in range(numCourses):
            if not self.dfs(c, set()):
                return []
        return self.order
    def dfs(self, c, vis):
        if c in self.order:
            return True
        if c in vis:
            return False
        vis.add(c)
        for d in self.adj[c]:
            if not self.dfs(d, vis):
                return False
        vis.remove(c)
        self.order.append(c)
        return True

